Limit authentication to specific users

sparkyb opened the pull request:
This PR adds support for limiting which authenticated users are allowed to perform each action. This enables you to password protect all downloads while being able to further limit which users are allowed to upload.

In addition to being able to list allowed users for each action, this also supports loading an Apache-like group file so that you can use groups to specify pre-defined sets of users.

My use case is that I want to have a private package repository for code that can't be available publicly. This means I want to limit listing and downloads to only members of my organization. However, I don't want everyone who has read access to also have write access. This PR allows me add as many users as I want to my password file and give them all read access, but specify only specific users who can upload (either by using a list of users on the command-line or via a group file).

I considered using separate files for each action, but since users who can write almost always should be able to read, I'd have to make sure those users are in both files with the same password. I previous did a version of this where I used Nginx to do the read authentication with one file, and pypiserver did authentication only for update with a different file.

I realize the command-line syntax is a little messy, but I wanted to preserve backwards compatibility. I'm willing to accept alternate suggestions. The way I have it now it is `-a <action>[=<user>[:<user>]...][,...]`. Still a comma-separated list of actions. An action by itself means any authenticated user is allowed, or an action can be followed by an = and a list of users. The users are separated with colons right now because comma was already used for separating actions. One possible alternative is to allow multiple `-a` arguments and have them use the `append` action so we could use comma as a user separator like `-a list -a download -a upload=user1,user2`. With this, we could still support comma-separated actions for backwards compatibility if there's no =, i.e. `-a list,download -a upload=user1,user2`. Another alternative is adding a new option for the new behavior and merging it with any legacy `-a` options, but what would it mean if an action were used in both?

Also, I used an @ prefix to indicate groups. I don't love this but there didn't seem to be an obvious symbol. I'm open to suggestions. devpi uses : but I couldn't do that if I'm colon-separating the users, but if I change it to the multi-argument style with commas, I could use a colon prefix for groups. Alternatively, I could use no prefix and treat every argument as possibly a group.

mplanchard commented:
Thanks for this PR! Just FYI I'm extremely busy right now w/work and a few things that have come up outside of work, so it will be probably a couple of weeks before I can get back to reviewing PRs. I will get to them as soon as possible!

munkim commented:
@mplanchard Any progress on this?

sparkyb commented:
@mplanchard Any update on when you might be able to review this and my other PR?

mplanchard commented:
Sorry for the delay, but I've largely stepped away from maintenance of
pypiserver. @dee-me-tree-or-love has taken up the mantle, but I know he
is also busy with school.

Ben Buchwald ***@***.***> writes:

> @mplanchard Any update on when you might be able to review this and my other PR?
>
> —
> Reply to this email directly, view it on GitHub, or unsubscribe.
> You are receiving this because you were mentioned.

sparkyb commented:
@dee-me-tree-or-love Nice work on getting 1.5.0 out. Any chance this and my other PR can get reviewed for possible inclusion into a next version?

dee-me-tree-or-love commented:
> @dee-me-tree-or-love Nice work on getting 1.5.0 out. Any chance this and my other PR can get reviewed for possible inclusion into a next version?

Hi @sparkyb, thanks a lot for reminding me and thanks for this PR! I'm sorry for the lack of response so far, like @mplanchard indicated, it's been quite a busy time :) Coming to these PRs, I will of course be happy to review them. All in all, I plan to make a selection of some features for the next minor release, so I'll review and aim to get them there too. I am still finishing a few responsibilities for this quartile, and it might take a bit, but I am aiming to get a new small version out somewhere by the end of next month 🤞 :) 

In the meanwhile, could you please update your branch with the `master` so that we run the CI/CD and all the automated checks? :) Thank you!

dee-me-tree-or-love commented:
Hey @sparkyb ! Thanks again for the PR, I really like the idea 👍 I will be looking into it in the coming days and will try to leave a review before I go on a small summer break. I know it's been quite a while since this was proposed, but would you be up for joining the discussion on this? :) I will start with trying this out locally and leaving the code-related notes, and then start thinking about the points you also bring up regarding the auth specification syntax. Probably I will release a small patch release before this is merged, but I'd be happy to aim for the next minor version. How does this sound? Will be happy to hear from you, cheers :) ✌️

sparkyb commented:
@dee-me-tree-or-love Sorry I didn't respond earlier. GitHub wasn't notifying me about comments, only reviews, so I fixed that now. Yes, I'm still interested in this and happy to discuss it if you like. I've been using my fork for over a year now and have been pretty happy with it, but would love to see it merged into the official version. I rebased onto the latest changes and it still seems good.

dee-me-tree-or-love commented:
Hey @sparkyb! New apologies for the new silence on my end! Don't worry about the responses, I'm happy to hear that this is still relevant and that you've been using it privately already 👍 I would like to make this a part of `pypiserver`, but I would like to aim for this as a milestone past a [release of `2.0.0`](https://github.com/pypiserver/pypiserver/milestone/11), so something like `2.1.0`. I'd like to focus on getting everything ready for `2.0.0` in the near future and would be interested to zoom into this. Let me know what you think! I'll then review this with full attention once the release groundwork is done and it's running neatly, then I'll also be more at ease with pushing some new nice features :)

sparkyb commented:
@dee-me-tree-or-love That sounds good to me. Let me know if there's anything else you need me to do to help.

sparkyb commented:
I've rebased this and #379 on top of master (2.x) so hopefully they can be included in a future release soon and I can stop maintaining my private fork.

dee-me-tree-or-love commented:
> I've rebased this and #379 on top of master (2.x) so hopefully they can be included in a future release soon and I can stop maintaining my private fork.

Hey @sparkyb! Thank you so much for all the updates and your efforts in maintaining the PR. I'll do my best to make it a part of the upcoming release. Thanks and sorry for all the waiting! Putting this on a list for my work in coming year.

Hope you have a good holiday season and a good start of 2024.
